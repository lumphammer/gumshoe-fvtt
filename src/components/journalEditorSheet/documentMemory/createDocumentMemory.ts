import { createStack } from "./createStack";
import { DocumentMemory } from "./types";

/**
 * Create a new empty document memory
 */
export function createDocumentMemory(
  period: number,
  maxDepth: number | null = null,
): DocumentMemory {
  return {
    stack: createStack(),
    serial: 0,
    period,
    maxDepth,
    state: "",
    snapshots: [""],
  };
}
