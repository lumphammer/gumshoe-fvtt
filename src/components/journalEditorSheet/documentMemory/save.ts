import createDiff from "textdiff-create";
import applyDiff from "textdiff-patch";

import { createStack } from "./createStack";
import { isMagicSerial } from "./isMagicSerial";
import { DocumentMemory, Edit, Stack } from "./types";

/**
 * This function pushes a new state onto a stack and returns the new stack.
 * Recurses down onto deeper stacks if necessary.
 */
function push(
  stack: Stack,
  newEdit: Edit,
  serial: number,
  period: number,
  depth: number,
  maxDepth: number | null,
  snapshots: string[],
): [Stack, string[]] {
  let nextStack = stack.next;

  // the new bombBay is the old bombBay plus the overflow from edits
  let bombBay: Edit[] = [...stack.bombBay];
  if (stack.edits.length >= period) {
    bombBay.push(stack.edits[0]);
  }

  // the new edits
  const edits = [
    ...stack.edits.slice(stack.edits.length === period ? 1 : 0),
    newEdit,
  ];

  let newSnapshots = snapshots;

  if (
    isMagicSerial(period, depth, serial) &&
    (maxDepth === null || depth < maxDepth)
  ) {
    // calculate the current snapshot by applying edits
    const [oldState = "", ...nextSnapshots] = snapshots;
    let newState = oldState;
    for (const edit of bombBay) {
      newState = applyDiff(newState, edit.changes);
    }
    const editToPush: Edit = {
      changes: createDiff(oldState, newState),
      timestamp: bombBay[bombBay.length - 1].timestamp,
      serial: bombBay[bombBay.length - 1].serial,
    };
    bombBay = [];

    let newNextSnapshots = [];
    [nextStack, newNextSnapshots] = push(
      nextStack || createStack(),
      editToPush,
      serial,
      period,
      depth + 1,
      maxDepth,
      nextSnapshots.length > 0 ? nextSnapshots : [""],
    );
    newSnapshots = [newState, ...newNextSnapshots];
  }

  return [
    {
      bombBay,
      edits,
      next: nextStack,
    },
    newSnapshots,
  ];
}

/**
 * Add a new state to the memory
 */
export function save(memory: DocumentMemory, state: string): DocumentMemory {
  if (state === memory.state) return memory;
  const serial = memory.serial + 1;
  const changes = createDiff(memory.state, state);
  const edit: Edit = {
    changes,
    timestamp: Math.floor(Date.now() / 1000),
    serial,
  };
  const [stack, snapshots] = push(
    memory.stack,
    edit,
    serial,
    memory.period,
    1,
    memory.maxDepth,
    memory.snapshots,
  );
  return {
    stack,
    serial,
    period: memory.period,
    maxDepth: memory.maxDepth,
    state,
    snapshots,
  };
}
