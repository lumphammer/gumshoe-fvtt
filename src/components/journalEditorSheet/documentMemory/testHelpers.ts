import { vi } from "vitest";

export const epoch = 0;

export function getAdditiveStates(count: number): string[] {
  if (count < 1 || count > 26) {
    throw new Error("count must be between 1 and 26");
  }
  const result: string[] = [];
  // loop over the indices of the final array
  for (let i = 0; i < count; i++) {
    const state = new Array(i + 1)
      .fill(null)
      .map((_, i) => String.fromCharCode(i + 97));
    result.push(state.join(""));
  }
  return result;
}

export function advanceTime10s(): void {
  vi.setSystemTime(new Date().setSeconds(new Date().getSeconds() + 10));
}
