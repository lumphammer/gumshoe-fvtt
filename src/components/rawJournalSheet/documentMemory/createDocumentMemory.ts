import { createStack } from "./createStack";
import { DocumentMemory } from "./types";

/**
 * Create a new empty document memory
 */
export function createDocumentMemory(periodicity: number): DocumentMemory {
  return {
    stack: createStack(periodicity),
    serial: 0,
    period: periodicity,
  };
}
