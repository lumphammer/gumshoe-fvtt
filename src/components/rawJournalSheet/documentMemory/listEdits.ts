import { DocumentMemory, Edit, Stack } from "./types";

function getStacks(memory: DocumentMemory): Stack[] {
  const stacks: Stack[] = [];
  for (let stack: Stack | null = memory.stack; stack; stack = stack.next) {
    stacks.push(stack);
  }
  return stacks;
}

export function listEdits(memory: DocumentMemory) {
  const stacks = getStacks(memory);
  const edits: Edit[] = [];
  let lastTimestamp = Infinity;
  for (const stack of stacks) {
    for (let i = stack.edits.length - 1; i >= 0; i--) {
      const edit = stack.edits[i];
      if (edit && edit.timestamp < lastTimestamp) {
        edits.push(edit);
        lastTimestamp = edit?.timestamp;
      }
    }
  }
  // yeah I could return edits.reverse(), but I hate the way it's side-effecty
  // so I like to treat it as if it doesn't return.
  edits.reverse();
  return edits;
}

// export function listEditsSince(memory: DocumentMemory, timestamp: number) {
//   const stacks = getStacks(memory);
//   const edits: Edit[] = [];
//   let lastTimestamp = Number.MAX_SAFE_INTEGER;
//   for (const stack of stacks) {
//     for (let i = stack.edits.length - 1; i >= 0; i--) {
//       const edit = stack.edits[i];
//       if (edit && edit.timestamp < lastTimestamp && edit.timestamp > timestamp) {
//         edits.push(edit);
//         lastTimestamp = edit?.timestamp ?? Infinity;
//       }
//     }
//   }
//   return edits;
// }
