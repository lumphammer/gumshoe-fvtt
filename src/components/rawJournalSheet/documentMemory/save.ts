import diff from "textdiff-create";

import { createStack } from "./createStack";
import { isMagicSerial } from "./isMagicSerial";
import { DocumentMemory, Edit, Stack } from "./types";

/**
 * This function pushes a new state onto a stack and returns the new stack.
 * Recurses down onto deeper stacks if necessary.
 */
function push(
  stack: Stack,
  newState: string,
  serial: number,
  period: number,
  depth: number,
): Stack {
  const newDelta: Edit = {
    change: diff(stack.snapshot, newState),
    timestamp: Math.floor(new Date().getTime() / 1000),
  };
  let next = stack.next;
  if (isMagicSerial(period, depth, serial)) {
    next = push(
      next ?? createStack(period),
      stack.snapshot,
      serial,
      period,
      depth + 1,
    );
  }
  return {
    edits: [...stack.edits.slice(1), newDelta],
    snapshot: newState,
    next,
  };
}

/**
 * Add a new state to the memory
 */
export function save(memory: DocumentMemory, state: string): DocumentMemory {
  const serial = memory.serial + 1;
  const stack = push(memory.stack, state, serial, memory.period, 1);
  return {
    stack,
    serial,
    period: memory.period,
  };
}
