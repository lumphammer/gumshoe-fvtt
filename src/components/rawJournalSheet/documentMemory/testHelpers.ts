import { vi } from "vitest";

export const epoch = 0;

const twentyStates = [
  "a",
  "ab",
  "abc",
  "abcd",
  "abcde",
  "abcdef",
  "abcdefg",
  "abcdefgh",
  "abcdefghi",
  "abcdefghij",
  "abcdefghijk",
  "abcdefghijkl",
  "abcdefghijklm",
  "abcdefghijklmn",
  "abcdefghijklmno",
  "abcdefghijklmnop",
  "abcdefghijklmnopq",
  "abcdefghijklmnopqr",
  "abcdefghijklmnopqrs",
  "abcdefghijklmnopqrst",
];

export const getStates = (n: number) => twentyStates.slice(0, n);

export function advanceTime10s() {
  vi.setSystemTime(new Date().setSeconds(new Date().getSeconds() + 10));
}
