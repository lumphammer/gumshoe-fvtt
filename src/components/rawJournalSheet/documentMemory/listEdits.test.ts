import { beforeEach, expect, test, vi } from "vitest";

import { createDocumentMemory } from "./createDocumentMemory";
import { listEdits } from "./listEdits";
import { save } from "./save";
import { advanceTime10s, epoch, getAdditiveStates } from "./testHelpers";

beforeEach(() => {
  vi.setSystemTime(new Date(epoch));
});

test.each(new Array(20).fill(null).map((_, i) => i))(
  "listEdits: %i edits",
  (i) => {
    let memory = createDocumentMemory(3);
    const states = getAdditiveStates(i);
    for (const state of states) {
      memory = save(memory, state);
      advanceTime10s();
    }
    const edits = listEdits(memory);
    expect(edits).toMatchSnapshot();
  },
);
