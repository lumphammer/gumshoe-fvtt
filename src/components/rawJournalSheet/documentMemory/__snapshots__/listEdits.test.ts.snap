// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`listEdits: 0 edits 1`] = `[]`;

exports[`listEdits: 1 edits 1`] = `
[
  {
    "change": [
      [
        1,
        "a",
      ],
    ],
    "timestamp": 0,
  },
]
`;

exports[`listEdits: 2 edits 1`] = `
[
  {
    "change": [
      [
        1,
        "a",
      ],
    ],
    "timestamp": 0,
  },
  {
    "change": [
      [
        0,
        1,
      ],
      [
        1,
        "b",
      ],
    ],
    "timestamp": 10,
  },
]
`;

exports[`listEdits: 3 edits 1`] = `
[
  {
    "change": [
      [
        1,
        "a",
      ],
    ],
    "timestamp": 0,
  },
  {
    "change": [
      [
        0,
        1,
      ],
      [
        1,
        "b",
      ],
    ],
    "timestamp": 10,
  },
  {
    "change": [
      [
        0,
        2,
      ],
      [
        1,
        "c",
      ],
    ],
    "timestamp": 20,
  },
]
`;

exports[`listEdits: 4 edits 1`] = `
[
  {
    "change": [
      [
        0,
        1,
      ],
      [
        1,
        "b",
      ],
    ],
    "timestamp": 10,
  },
  {
    "change": [
      [
        0,
        2,
      ],
      [
        1,
        "c",
      ],
    ],
    "timestamp": 20,
  },
  {
    "change": [
      [
        0,
        3,
      ],
      [
        1,
        "d",
      ],
    ],
    "timestamp": 30,
  },
]
`;

exports[`listEdits: 5 edits 1`] = `
[
  {
    "change": [
      [
        0,
        2,
      ],
      [
        1,
        "c",
      ],
    ],
    "timestamp": 20,
  },
  {
    "change": [
      [
        0,
        3,
      ],
      [
        1,
        "d",
      ],
    ],
    "timestamp": 30,
  },
  {
    "change": [
      [
        0,
        4,
      ],
      [
        1,
        "e",
      ],
    ],
    "timestamp": 40,
  },
]
`;

exports[`listEdits: 6 edits 1`] = `
[
  {
    "change": [
      [
        1,
        "abc",
      ],
    ],
    "timestamp": 20,
  },
  {
    "change": [
      [
        0,
        3,
      ],
      [
        1,
        "d",
      ],
    ],
    "timestamp": 30,
  },
  {
    "change": [
      [
        0,
        4,
      ],
      [
        1,
        "e",
      ],
    ],
    "timestamp": 40,
  },
  {
    "change": [
      [
        0,
        5,
      ],
      [
        1,
        "f",
      ],
    ],
    "timestamp": 50,
  },
]
`;

exports[`listEdits: 7 edits 1`] = `
[
  {
    "change": [
      [
        1,
        "abc",
      ],
    ],
    "timestamp": 20,
  },
  {
    "change": [
      [
        0,
        4,
      ],
      [
        1,
        "e",
      ],
    ],
    "timestamp": 40,
  },
  {
    "change": [
      [
        0,
        5,
      ],
      [
        1,
        "f",
      ],
    ],
    "timestamp": 50,
  },
  {
    "change": [
      [
        0,
        6,
      ],
      [
        1,
        "g",
      ],
    ],
    "timestamp": 60,
  },
]
`;

exports[`listEdits: 8 edits 1`] = `
[
  {
    "change": [
      [
        1,
        "abc",
      ],
    ],
    "timestamp": 20,
  },
  {
    "change": [
      [
        0,
        5,
      ],
      [
        1,
        "f",
      ],
    ],
    "timestamp": 50,
  },
  {
    "change": [
      [
        0,
        6,
      ],
      [
        1,
        "g",
      ],
    ],
    "timestamp": 60,
  },
  {
    "change": [
      [
        0,
        7,
      ],
      [
        1,
        "h",
      ],
    ],
    "timestamp": 70,
  },
]
`;

exports[`listEdits: 9 edits 1`] = `
[
  {
    "change": [
      [
        1,
        "abc",
      ],
    ],
    "timestamp": 20,
  },
  {
    "change": [
      [
        0,
        3,
      ],
      [
        1,
        "def",
      ],
    ],
    "timestamp": 50,
  },
  {
    "change": [
      [
        0,
        6,
      ],
      [
        1,
        "g",
      ],
    ],
    "timestamp": 60,
  },
  {
    "change": [
      [
        0,
        7,
      ],
      [
        1,
        "h",
      ],
    ],
    "timestamp": 70,
  },
  {
    "change": [
      [
        0,
        8,
      ],
      [
        1,
        "i",
      ],
    ],
    "timestamp": 80,
  },
]
`;

exports[`listEdits: 10 edits 1`] = `
[
  {
    "change": [
      [
        1,
        "abc",
      ],
    ],
    "timestamp": 20,
  },
  {
    "change": [
      [
        0,
        3,
      ],
      [
        1,
        "def",
      ],
    ],
    "timestamp": 50,
  },
  {
    "change": [
      [
        0,
        7,
      ],
      [
        1,
        "h",
      ],
    ],
    "timestamp": 70,
  },
  {
    "change": [
      [
        0,
        8,
      ],
      [
        1,
        "i",
      ],
    ],
    "timestamp": 80,
  },
  {
    "change": [
      [
        0,
        9,
      ],
      [
        1,
        "j",
      ],
    ],
    "timestamp": 90,
  },
]
`;

exports[`listEdits: 11 edits 1`] = `
[
  {
    "change": [
      [
        1,
        "abc",
      ],
    ],
    "timestamp": 20,
  },
  {
    "change": [
      [
        0,
        3,
      ],
      [
        1,
        "def",
      ],
    ],
    "timestamp": 50,
  },
  {
    "change": [
      [
        0,
        8,
      ],
      [
        1,
        "i",
      ],
    ],
    "timestamp": 80,
  },
  {
    "change": [
      [
        0,
        9,
      ],
      [
        1,
        "j",
      ],
    ],
    "timestamp": 90,
  },
  {
    "change": [
      [
        0,
        10,
      ],
      [
        1,
        "k",
      ],
    ],
    "timestamp": 100,
  },
]
`;

exports[`listEdits: 12 edits 1`] = `
[
  {
    "change": [
      [
        1,
        "abc",
      ],
    ],
    "timestamp": 20,
  },
  {
    "change": [
      [
        0,
        3,
      ],
      [
        1,
        "def",
      ],
    ],
    "timestamp": 50,
  },
  {
    "change": [
      [
        0,
        6,
      ],
      [
        1,
        "ghi",
      ],
    ],
    "timestamp": 80,
  },
  {
    "change": [
      [
        0,
        9,
      ],
      [
        1,
        "j",
      ],
    ],
    "timestamp": 90,
  },
  {
    "change": [
      [
        0,
        10,
      ],
      [
        1,
        "k",
      ],
    ],
    "timestamp": 100,
  },
  {
    "change": [
      [
        0,
        11,
      ],
      [
        1,
        "l",
      ],
    ],
    "timestamp": 110,
  },
]
`;

exports[`listEdits: 13 edits 1`] = `
[
  {
    "change": [
      [
        0,
        3,
      ],
      [
        1,
        "def",
      ],
    ],
    "timestamp": 50,
  },
  {
    "change": [
      [
        0,
        6,
      ],
      [
        1,
        "ghi",
      ],
    ],
    "timestamp": 80,
  },
  {
    "change": [
      [
        0,
        10,
      ],
      [
        1,
        "k",
      ],
    ],
    "timestamp": 100,
  },
  {
    "change": [
      [
        0,
        11,
      ],
      [
        1,
        "l",
      ],
    ],
    "timestamp": 110,
  },
  {
    "change": [
      [
        0,
        12,
      ],
      [
        1,
        "m",
      ],
    ],
    "timestamp": 120,
  },
]
`;

exports[`listEdits: 14 edits 1`] = `
[
  {
    "change": [
      [
        0,
        3,
      ],
      [
        1,
        "def",
      ],
    ],
    "timestamp": 50,
  },
  {
    "change": [
      [
        0,
        6,
      ],
      [
        1,
        "ghi",
      ],
    ],
    "timestamp": 80,
  },
  {
    "change": [
      [
        0,
        11,
      ],
      [
        1,
        "l",
      ],
    ],
    "timestamp": 110,
  },
  {
    "change": [
      [
        0,
        12,
      ],
      [
        1,
        "m",
      ],
    ],
    "timestamp": 120,
  },
  {
    "change": [
      [
        0,
        13,
      ],
      [
        1,
        "n",
      ],
    ],
    "timestamp": 130,
  },
]
`;

exports[`listEdits: 15 edits 1`] = `
[
  {
    "change": [
      [
        0,
        3,
      ],
      [
        1,
        "def",
      ],
    ],
    "timestamp": 50,
  },
  {
    "change": [
      [
        0,
        6,
      ],
      [
        1,
        "ghi",
      ],
    ],
    "timestamp": 80,
  },
  {
    "change": [
      [
        0,
        9,
      ],
      [
        1,
        "jkl",
      ],
    ],
    "timestamp": 110,
  },
  {
    "change": [
      [
        0,
        12,
      ],
      [
        1,
        "m",
      ],
    ],
    "timestamp": 120,
  },
  {
    "change": [
      [
        0,
        13,
      ],
      [
        1,
        "n",
      ],
    ],
    "timestamp": 130,
  },
  {
    "change": [
      [
        0,
        14,
      ],
      [
        1,
        "o",
      ],
    ],
    "timestamp": 140,
  },
]
`;

exports[`listEdits: 16 edits 1`] = `
[
  {
    "change": [
      [
        0,
        6,
      ],
      [
        1,
        "ghi",
      ],
    ],
    "timestamp": 80,
  },
  {
    "change": [
      [
        0,
        9,
      ],
      [
        1,
        "jkl",
      ],
    ],
    "timestamp": 110,
  },
  {
    "change": [
      [
        0,
        13,
      ],
      [
        1,
        "n",
      ],
    ],
    "timestamp": 130,
  },
  {
    "change": [
      [
        0,
        14,
      ],
      [
        1,
        "o",
      ],
    ],
    "timestamp": 140,
  },
  {
    "change": [
      [
        0,
        15,
      ],
      [
        1,
        "p",
      ],
    ],
    "timestamp": 150,
  },
]
`;

exports[`listEdits: 17 edits 1`] = `
[
  {
    "change": [
      [
        0,
        6,
      ],
      [
        1,
        "ghi",
      ],
    ],
    "timestamp": 80,
  },
  {
    "change": [
      [
        0,
        9,
      ],
      [
        1,
        "jkl",
      ],
    ],
    "timestamp": 110,
  },
  {
    "change": [
      [
        0,
        14,
      ],
      [
        1,
        "o",
      ],
    ],
    "timestamp": 140,
  },
  {
    "change": [
      [
        0,
        15,
      ],
      [
        1,
        "p",
      ],
    ],
    "timestamp": 150,
  },
  {
    "change": [
      [
        0,
        16,
      ],
      [
        1,
        "q",
      ],
    ],
    "timestamp": 160,
  },
]
`;

exports[`listEdits: 18 edits 1`] = `
[
  {
    "change": [
      [
        0,
        6,
      ],
      [
        1,
        "ghi",
      ],
    ],
    "timestamp": 80,
  },
  {
    "change": [
      [
        0,
        9,
      ],
      [
        1,
        "jkl",
      ],
    ],
    "timestamp": 110,
  },
  {
    "change": [
      [
        0,
        12,
      ],
      [
        1,
        "mno",
      ],
    ],
    "timestamp": 140,
  },
  {
    "change": [
      [
        0,
        15,
      ],
      [
        1,
        "p",
      ],
    ],
    "timestamp": 150,
  },
  {
    "change": [
      [
        0,
        16,
      ],
      [
        1,
        "q",
      ],
    ],
    "timestamp": 160,
  },
  {
    "change": [
      [
        0,
        17,
      ],
      [
        1,
        "r",
      ],
    ],
    "timestamp": 170,
  },
]
`;

exports[`listEdits: 19 edits 1`] = `
[
  {
    "change": [
      [
        1,
        "abcdefghi",
      ],
    ],
    "timestamp": 80,
  },
  {
    "change": [
      [
        0,
        9,
      ],
      [
        1,
        "jkl",
      ],
    ],
    "timestamp": 110,
  },
  {
    "change": [
      [
        0,
        12,
      ],
      [
        1,
        "mno",
      ],
    ],
    "timestamp": 140,
  },
  {
    "change": [
      [
        0,
        16,
      ],
      [
        1,
        "q",
      ],
    ],
    "timestamp": 160,
  },
  {
    "change": [
      [
        0,
        17,
      ],
      [
        1,
        "r",
      ],
    ],
    "timestamp": 170,
  },
  {
    "change": [
      [
        0,
        18,
      ],
      [
        1,
        "s",
      ],
    ],
    "timestamp": 180,
  },
]
`;
